internal/locate/region_request.go

```diff
@@ -262,112 +481,184 @@ func newReplicaSelector(regionCache *RegionCache, regionID RegionVerID) (*replic
+		if s.replicas[newLeaderIdx].attempts == maxReplicaAttempt {
+			s.replicas[newLeaderIdx].attempts--
 		}
-		replica.attempts++
+		return
+	}
+}
 
-		storeFailEpoch := atomic.LoadUint32(&replica.store.epoch)
-		if storeFailEpoch != replica.epoch {
-			// TODO(youjiali1995): Is it necessary to invalidate the region?
-			metrics.TiKVReplicaSelectorFailureCounter.WithLabelValues("stale_store").Inc()
-			s.invalidateRegion()
-			return nil, nil
+func (s *replicaSelector) buildRPCContext(bo *retry.Backoffer) (*RPCContext, error) {
+	targetReplica, proxyReplica := s.targetReplica(), s.proxyReplica()
+
+	// Backoff and retry if no replica is selected or the selected replica is stale
+	if targetReplica == nil || s.isReplicaStoreEpochStale(targetReplica) ||
+		(proxyReplica != nil && s.isReplicaStoreEpochStale(proxyReplica)) {
+		return nil, nil
+	}
+
+	rpcCtx := &RPCContext{
+		Region:     s.region.VerID(),
+		Meta:       s.region.meta,
+		Peer:       targetReplica.peer,
+		Store:      targetReplica.store,
+		AccessMode: tiKVOnly,
+		TiKVNum:    len(s.replicas),
+	}
+
+	// Set leader addr
+	addr, err := s.regionCache.getStoreAddr(bo, s.region, targetReplica.store)
+	if err != nil {
+		return nil, err
+	}
+	if len(addr) == 0 {
+		return nil, nil
+	}
+	rpcCtx.Addr = addr
+	targetReplica.attempts++
+
+	// Set proxy addr
+	if proxyReplica != nil {
+		addr, err = s.regionCache.getStoreAddr(bo, s.region, proxyReplica.store)
+		if err != nil {
+			return nil, err
 		}
-		addr, err := s.regionCache.getStoreAddr(bo, s.region, replica.store)
-		if err == nil && len(addr) != 0 {
-			return &RPCContext{
-				Region:     s.region.VerID(),
-				Meta:       s.region.meta,
-				Peer:       replica.peer,
-				Store:      replica.store,
-				Addr:       addr,
-				AccessMode: tiKVOnly,
-				TiKVNum:    len(s.replicas),
-			}, nil
+		if len(addr) == 0 {
+			return nil, nil
 		}
+		rpcCtx.ProxyStore = proxyReplica.store
+		rpcCtx.ProxyAddr = addr
+		proxyReplica.attempts++
 	}
+
+	return rpcCtx, nil
+}
+
+func (s *replicaSelector) isReplicaStoreEpochStale(replica *replica) bool {
+	storeFailEpoch := atomic.LoadUint32(&replica.store.epoch)
+	if storeFailEpoch != replica.epoch {
+		// TODO(youjiali1995): Is it necessary to invalidate the region?
+		metrics.TiKVReplicaSelectorFailureCounter.WithLabelValues("stale_store").Inc()
+		s.invalidateRegion()
+		return true
+	}
+	return false
 }
 
 func (s *replicaSelector) onSendFailure(bo *retry.Backoffer, err error) {
 	metrics.RegionCacheCounterWithSendFail.Inc()
-	replica := s.replicas[s.nextReplicaIdx-1]
-	if replica.store.requestLiveness(bo, s.regionCache) == reachable {
-		s.rewind()
-		return
+	s.state.onSendFailure(bo, s, err)
+}
+
+func (s *replicaSelector) checkLiveness(bo *retry.Backoffer, accessReplica *replica) livenessState {
+	store := accessReplica.store
+	liveness := store.requestLiveness(bo, s.regionCache)
+	if liveness != reachable {
+		store.startHealthCheckLoopIfNeeded(s.regionCache)
```

Comment:
Yeah.. I hope to maintain the liveness state of all stores. Even when forwarding is not enabled, we can use this to decide what to do, like directly try other replicas or return error early to the user (not implemented in this PR).

Reply:
I'm afraid it might not works very well. The check loop is very roughly implemented. It creates one goroutine for each unreachable nodes. If you are going to use it no matter if the forwarding feature is enabled, I think you might need to give it a better implementation. 😥

Reply:
Ok, at now I change to only start this when forwarding is enabled. If necessary, we can improve it and check health in background.